nav2_collision_monitor/src/collision_monitor_node.cpp

```diff
@@ -0,0 +1,530 @@
+          node, tf_buffer_, source_name, base_frame_id,
+          transform_tolerance, data_timeout_);
+
+        p->configure();
+
+        sources_.push_back(p);
+      } else {  // Error if something else
+        RCLCPP_ERROR(
+          get_logger(),
+          "[%s]: Unknown source type: %s",
+          source_name.c_str(), source_type.c_str());
+        return false;
+      }
+    }
+  } catch (const std::exception & ex) {
+    RCLCPP_ERROR(get_logger(), "Error while getting parameters: %s", ex.what());
+    return false;
+  }
+
+  return true;
+}
+
+void CollisionMonitor::processMain()
+{
+  // Current timestamp for all inner routines prolongation
+  rclcpp::Time curr_time = this->now();
+
+  // Do nothing if main worker in non-active state
+  // or if velocity is not received yet
+  if (!process_active_ || !velocityValid(curr_time)) {
+    return;
+  }
+
+  // Points array collected from different data sources in a robot base frame
+  std::vector<Point> collision_points;
+
+  // Fill collision_points array from different data sources
+  for (std::shared_ptr<SourceBase> source : sources_) {
+    source->getData(curr_time, collision_points);
+  }
+
+  // By default - there is no action
+  Action robot_action{DO_NOTHING, {0.0, 0.0, 0.0}};
+  // Used for APPROACH model to define "nearest" obstacle
+  double min_collision_time = std::numeric_limits<double>::max();
+
+  for (std::shared_ptr<PolygonBase> polygon : polygons_) {
+    const ActionType at = polygon->getActionType();
+    if (robot_action.action_type == STOP) {
+      // If robot already should stop, do nothing
+      break;
+    }
+
+    // Process selected polygon
+    if (at == STOP || at == SLOWDOWN) {
+      processStopSlowdown(polygon, collision_points, cmd_vel_in_, robot_action);
+    } else if (at == APPROACH) {
+      processApproach(polygon, collision_points, velocity_, min_collision_time, robot_action);
+    }
+  }
+
+  // In APPROACH model due to sensors noise calculated speed of robot might be instantly changed
+  // causing robot to twitch. To avoid this, when releasing robot after APPROACH,
+  // its speed should be increased gradually smoothing the twitches over time until robot will
+  // reach its desired normal opration.
+  if (robot_action.action_type == DO_NOTHING) {
+    if (robot_action_prev_.action_type == APPROACH) {
+      // Trigger gradually velocity increase
+      release_operation_ = true;
+    }
+
+    if (!release_operation_) {
+      // Continue normal robot operation
+      robot_action.req_vel = cmd_vel_in_;
+    } else {
+      // Release robot from approach: accelerate the robot (without changing speed instantly)
+      // in order to avoid twitching
+      releaseOperation(curr_time, cmd_vel_in_, robot_action);
+    }
+  }
+
+  if (robot_action != robot_action_prev_) {
+    // Report robot action type and changed velocity
+    printAction(robot_action);
+
+    // Publish required robot velocity
+    publishVelocity(robot_action.req_vel);
```

Comment:
You CANNOT only publish velocities when there is a change, this MUST be a constant stream unless the robot is stopped and the commands are `0`s

Reply:
OK, I think more: we also should continue to publish even zero-velocities, for the safety. E.g. if for some reasons first zero `cmd_vel_out` message was lost for some reasons (e.g. network issue between remote robot and brain server).

Reply:
It would not be necessary and probably not good to send `0`s if we're not receiving anything. It could be overriding other systems trying to talk to the robot base outside of Nav2. Plus if you send commands, a brushless motor can't go into standby mode to save energy. We should not publish after some timeout of no longer receiving updates from cmd_vel topic

Reply:
Okay, I see. Will make it in the next update.

Reply:
Now, publishVelocity() [has a logic](https://github.com/ros-planning/navigation2/pull/2982/files#diff-e1c76d88179c14a1cf399e94bebd4f5d17dedca39e09b3bbdc810aa11a705543R157) for 0-velocity ceasing to publish after `stop_pub_timeout` (set by developer).